bulbs/special_coverage/search.py

```diff
@@ -1,27 +1,122 @@
+        should_filters = [
+            es_filter.Terms(pk=self.query.get("included_ids", [])),
+            es_filter.Terms(pk=self.query.get("pinned_ids", []))
+        ]
+        should_filters += self.get_group_filters()
+
+        # Compile list of all Must filters.
+        must_filters = [
+            es_filter.Bool(should=should_filters),
+            ~es_filter.Terms(pk=self.query.get("excluded_ids", []))
+        ]
+
+        return Content.search_objects.search().filter(es_filter.Bool(must=must_filters))
+
+    def get_group_filters(self):
+        """Return es OR filters to include all special coverage group conditions."""
+        group_filters = []
+        field_map = {
+            "feature-type": "feature_type.slug",
+            "tag": "tags.slug",
+            "content-type": "_type"
+        }
+        for group_set in self.query.get("groups", []):
+            for group in group_set:
+                group_filter = es_filter.MatchAll()
+                for condition in group.get("conditions", []):
+                    group_filter &= get_condition_filter(condition, field_map=field_map)
+                group_filters.append(group_filter)
+        return group_filters
+
+    @property
+    def query(self):
+        """Group the self.special_coverages queries and memoize them."""
+        if not self._query:
+            self._query.update({
+                "excluded_ids": [],
+                "included_ids": [],
+                "pinned_ids": [],
+                "groups": [],
+            })
+            for special_coverage in self._special_coverages:
+                # Access query at dict level.
+                query = getattr(special_coverage, "query", {})
+                if "query" in query:
+                    query = query.get("query")
+                self._query["excluded_ids"] += query.get("excluded_ids", [])
+                self._query["included_ids"] += query.get("included_ids", [])
+                self._query["pinned_ids"] += query.get("pinned_ids", [])
+                self._query["groups"] += [query.get("groups", [])]
+        return self._query
+
+
+class SearchSlicer(object):
+    """We want to search things like a seesaw. take that mike parent.
     """
-    index = 0
-    while True:
-        result = None
-        if index == 1:
-            try:
-                result = six.next(query2)
-            except IndexError:
-                pass
-        if result is None:
-            try:
-                result = six.next(query1)
-            except IndexError:
-                break
-        yield result
-        index += 1
+    def __init__(self, *args, **kwargs):
+        self.default_queryset = None
+        self.querysets = OrderedDict()
+        self.index = 0
+        self.limit = kwargs.get("limit", getattr(settings, "READING_LIST_LIMIT", 100))
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        if self.index > self.limit:
```

Comment:
shouldn't this be `>=`?
If limit == 1, this won't be TRUE until we've returned 2 items (and index == 2)